### PR TITLE
Adding patch functionality for px

### PR DESCRIPTION
**Pull request by prashanthpx**
Part of #18 
patch is used to update volume parameters

1. Added new cobra command for "patch"
2.Supporting the following patching parameters: halevel, size, nodes, shared and sticky

3. Added UT's for patch
4. Helper function to validate volumeSpec Message. This is needed as some combination of patch parameters are not supported in portworx.

### CODE REVIEW

File: cmd/common_test.go [hunk 1]
```diff
@@ -53,6 +53,7 @@ func testCreateVolume(t *testing.T, volName string, size uint64) string {
 func testGetVolume(t *testing.T, volId string, volName string) bool {
 	cli := "px get volume -o wide " + volId
 	lines := executeCli(cli)
+	fmt.Printf(" get vol : %s", lines)
```
Comment (lpabon) [inline cmd/common_test.go]: Please remove these

**Comment by prashanthpx** [comment]
This PR is WIP due to dependency on other changes coming in. Please do not review now.

File: cmd/patchVolume.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	// for setting volume to be sticky
+	if updateReq.sticky {
+		updateReq.req.Spec.StickyOpt = &api.VolumeSpecUpdate_Sticky{
+			Sticky: updateReq.sticky,
+		}
+		ok = true
+	}
+
+	if !ok {
+		return util.PxErrorMessage(err, "Must supply any one of the flags with valid paramateres."+
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed.

File: cmd/patchVolume.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			"Please see help for more info")
+	}
+
+	// Before submitting the request, need to make sure volumespec is checked for valid flag combinations.
+	ret, err := portworx.ValidateVolumeSpec(updateReq.req.Spec)
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed it.

File: pkg/portworx/volspecvalidation.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	api "github.com/libopenstorage/openstorage-sdk-clients/sdk/golang"
+)
+
+/*
+Currently due to volume driver limitation, only few combination of fields can be
+set as paramaters in VolumeSpecUpdate. This function will check for the same.
+It is recommeneded every consumer of VolumeSpecUpdate needs to call this function
+to make sure non-supported flagh comibnation are errored out to user.
+*/
+func ValidateVolumeSpec(volspec *api.VolumeSpecUpdate) (bool, error) {
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply pkg/portworx/volspecvalidation.go]: Keeping scope for expansion, today it's only one case but in future, we might have many more combinations which might not be supported and which needs to be added.  The error message provides a scope to display an appropriate message based on the combination.

File: pkg/portworx/volspecvalidation.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	// case of checking possible halevel flag combination
+	if volspec.GetHaLevel() > 0 {
+		if volspec.GetSize() > 0 || volspec.GetShared() || volspec.GetSticky() {
+			// Please have unique msgs for each case so it's easy for use to identity the
+			// flags mismatch combination.
+			return false, fmt.Errorf("Invalid halevel flag combination. Size, Shared or Sticky flag not supported" +
+				"with halevel flag")
+		}
+	}
+	return true, fmt.Errorf("valid combination")
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply pkg/portworx/volspecvalidation.go]: Fixed it. Added nil.

File: pkg/portworx/volspecvalidation.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+*/
+package portworx
+
+import (
+	"fmt"
+
+	api "github.com/libopenstorage/openstorage-sdk-clients/sdk/golang"
+)
+
+/*
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply pkg/portworx/volspecvalidation.go]: Fixed it.

File: cmd/patchVolume_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	volCreate(t, volName)
+	// Now update shared to true
+	testPatchVolumeShared(t, volName, sharedTrue)
+	//Now unset shared aka to false
+	testPatchVolumeShared(t, volName, sharedFalse)
+	volCleanup(t, volName)
+}
+
+/*
+Helper to create a volume
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume_test.go]: Fixed it.

File: cmd/patchVolume.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+			return fmt.Errorf("Must supply a name for volume")
+		}
+		return nil
+	},
+	RunE: updateVolume,
+}
+
+func init() {
+	patchCmd.AddCommand(patchVolumeCmd)
+	patchVolumeCmd.Flags().Int64Var(&updateReq.halevel, "halevel", 0, "New replication factor (Valid Range: [1 3]) (default 1)")
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed it.

File: cmd/patchVolume.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	defer conn.Close()
+	// fetch the volume name from args
+	updateReq.req.VolumeId = args[0]
+
+	updateReq.req.Spec = &api.VolumeSpecUpdate{}
+	// By default blank replicaset needs to be passed due to know volume driver issuw (pwx-9500)
+	// If user provides one it will be overriden
+	updateReq.req.Spec.ReplicaSet = &api.ReplicaSet{
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed it.

File: cmd/patchVolume.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+// updateVolumeCmd represents the updateVolume command
+var patchVolumeCmd = &cobra.Command{
+	Use:   "volume [NAME]",
+	Short: "Update field(s) of a portworx volume",
+	Example: `$ px patch  volume test1 --halevel 3
+This set halevel of volume to 3.`,
+
+	Args: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 1 {
+			return fmt.Errorf("Must supply a name for volume")
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed it.

File: pkg/portworx/volspecvalidation.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+It is recommeneded every consumer of VolumeSpecUpdate needs to call this function
+to make sure non-supported flagh comibnation are errored out to user.
+*/
+func ValidateVolumeSpec(volspec *api.VolumeSpecUpdate) (bool, error) {
+	// case of checking possible halevel flag combination
+	if volspec.GetHaLevel() > 0 {
+		if volspec.GetSize() > 0 || volspec.GetShared() || volspec.GetSticky() {
+			// Please have unique msgs for each case so it's easy for use to identity the
+			// flags mismatch combination.
+			return false, fmt.Errorf("Invalid halevel flag combination. Size, Shared or Sticky flag not supported" +
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply pkg/portworx/volspecvalidation.go]: Fixed it.

File: cmd/common_test.go [hunk 12]
```diff
@@ -178,3 +178,27 @@ func testCreateClone(t *testing.T, volId string, cloneName string) {
 
 	assert.True(t, util.ListContainsSubString(lines, fmt.Sprintf("Clone of %s created with id", volId)))
 }
+
+func testPatchVolumeHalevel(t *testing.T, volName string, haLevel int) {
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/common_test.go]: Good suggestion. Made it as table-driven, fixed it.

**Comment by prashanthpx** [comment]
@ggriffiths - Addressed all review comments. Please review.

**Comment by ggriffiths** [comment]
Looks good! Just make sure to update your branch with the latest.

File: cmd/patchVolume_test.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+//Helper to create a volume
+func volCreate(t *testing.T, volName string) {
+	// Create a volume
+	testCreateVolume(t, volName, 1)
+	// Verify that the volume got created
+	assert.True(t, testHasVolume(volName))
+}
+
+//Helper function to cleanup volume created
```
Comment (siva-portworx) [inline cmd/patchVolume_test.go]: Minor: It is better to leave space after //.

File: cmd/patchVolume_test.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cmd
+
+import (
+	_ "fmt"
```
Comment (siva-portworx) [inline cmd/patchVolume_test.go]: We can remove "fmt" import as it is not used in the file.

File: cmd/patchVolume_test.go (hunk 13 above)
Comment (prashanthpx) [reply cmd/patchVolume_test.go]: Fixed.

File: cmd/patchVolume_test.go (hunk 14 above)
Comment (prashanthpx) [reply cmd/patchVolume_test.go]: Fixed.

File: cmd/patchVolume.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+		},
+		RunE: updateVolume,
+	}
+})
+
+var _ = RegisterCommandInit(func() {
+	patchCmd.AddCommand(patchVolumeCmd)
+
+	patchVolumeCmd.Flags().Int64Var(&updateReq.halevel, "halevel", 0, "New replication factor (Valid Range: [1, 3]) (default 1)")
+	patchVolumeCmd.Flags().StringSliceVarP(&updateReq.replicaSet, "nodes", "n", []string{}, "")
```
Comment (lpabon) [inline cmd/patchVolume.go]: Needs a help message

File: cmd/patchVolume.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	patchVolumeCmd.Flags().Uint64VarP(&updateReq.size, "size", "s", 0, "New size for the volume (GiB) (default 1)")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "shared", "r", false, "Shared is true if this volume can be remotely accessed [default false].")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "sticky", "t", false, "set sticky setting [default false].")
+	patchVolumeCmd.Flags().SortFlags = false
+})
+
+func updateVolume(cmd *cobra.Command, args []string) error {
+	ctx, conn, err := PxConnectDefault()
+	ok := false
+	if err != nil {
```
Comment (lpabon) [inline cmd/patchVolume.go]: This needs to be below line 75

File: cmd/patchVolume.go [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+	patchVolumeCmd.Flags().StringSliceVarP(&updateReq.replicaSet, "nodes", "n", []string{}, "")
+	patchVolumeCmd.Flags().Uint64VarP(&updateReq.size, "size", "s", 0, "New size for the volume (GiB) (default 1)")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "shared", "r", false, "Shared is true if this volume can be remotely accessed [default false].")
+	patchVolumeCmd.Flags().BoolVarP(&updateReq.shared, "sticky", "t", false, "set sticky setting [default false].")
+	patchVolumeCmd.Flags().SortFlags = false
+})
+
+func updateVolume(cmd *cobra.Command, args []string) error {
+	ctx, conn, err := PxConnectDefault()
+	ok := false
```
Comment (lpabon) [inline cmd/patchVolume.go]: This needs to be closer to the code that uses it, which seems to be line 92. Please add a comment as to why it is needed or change the variable name to be more descriptive.

File: pkg/portworx/volspecvalidation.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lpabon) [reply pkg/portworx/volspecvalidation.go]: I'm not sure I understand the reasoning. I agree with @ggriffiths in the this should just return an error where the function describe why it was not valid. Returning nil means the volspec is valid.

File: cmd/patchVolume.go [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+		return err
+	}
+
+	defer conn.Close()
+	// fetch the volume name from args
+	updateReq.req.VolumeId = args[0]
+
+	// By default blank replicaset needs to be passed due to know volume driver issuw (pwx-9500)
+	// If user provides one it will be overriden
+	updateReq.req.Spec = &api.VolumeSpecUpdate{
```
Comment (lpabon) [inline cmd/patchVolume.go]: There should be an `if` before this line, if there `len (updateReq.replicaSet) != 0`

File: cmd/patchVolume.go [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+
+	// for setting volume to be sticky
+	if updateReq.sticky {
+		updateReq.req.Spec.StickyOpt = &api.VolumeSpecUpdate_Sticky{
+			Sticky: updateReq.sticky,
+		}
+		ok = true
+	}
+
+	if !ok {
```
Comment (lpabon) [inline cmd/patchVolume.go]: Suggestion: maybe use reflect.DeepEqual and compare updateReq.req.Spec with an empty updateReq.req.Spec. If it is the same, then no settings were set.

File: cmd/common_test.go [hunk 20]
```diff
@@ -174,3 +174,30 @@ func testCreateClone(t *testing.T, volId string, cloneName string) {
 
 	assert.True(t, util.ListContainsSubString(lines, fmt.Sprintf("Clone of %s created with id", volId)))
 }
+
+func testPatchVolume(t *testing.T, volName string, haLevel int, size uint64, shared bool) {
```
[the remark this answers is not in the document]
Comment (lpabon) [reply cmd/common_test.go]: For example, a resize function in tests can be added in common_tests.go if necessary.

[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/common_test.go]: The first commit was that way, then changed it. Fixed it by adding small utility function and moving tests to patchVolume_test.go

File: cmd/patch.go [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/spf13/cobra"
+)
+
+var patchCmd *cobra.Command
+
+// patchCmd represents the patch command
+var _ = RegisterCommandVar(func() {
+	patchCmd = &cobra.Command{
+		Use:   "patch",
+		Short: "Update field(s) an object in Portworx",
```
[the remark this answers is not in the document]
Comment (prashanthpx) [reply cmd/patch.go]: Fixed it.

File: cmd/patchVolume.go (hunk 15 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: My bad overlooked it. Fixed it.

File: cmd/patchVolume.go (hunk 16 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed it.

File: cmd/patchVolume.go (hunk 18 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: Reason for not having an if() check is "replicaSet" always needs to be passed or leads to PWX-9500 issue.  So either it's empty or user-provided, by default it's empty.

File: cmd/patchVolume.go (hunk 19 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: It's not exactly empty struct but with replicaset filled (default value is empty). Sounds good, but we would need a deepcopy library vendored in. Can I use this one ```import "github.com/mohae/deepcopy"```

File: cmd/patchVolume.go (hunk 17 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: Fixed, added a descriptive variable name.

File: cmd/patchVolume.go (hunk 18 above)
Comment (prashanthpx) [reply cmd/patchVolume.go]: As discussed passing replicaSet only if halevel is set.

**Pull request merged**